### DISCUSSION TITLE
Lualine, Noice, and Macro Recording

### DISCUSSION THREAD

**Discussion by lplawlor** [nix-community/nixvim#1653] [Q&A] [answered]
With lualine and noice enabled in their default states, you can no longer see when you are recording a new macro with `q`.

There's a section of the Noice docs that discusses how to add back this functionality via either lualine or a notify message:
https://github.com/folke/noice.nvim/wiki/Configuration-Recipes#show-recording-messages

Would anyone know how to translate these solutions to a Nixvim configuration?

Edit: I got the notify solution working by doing the following:
```nix
programs.nixvim = {
  enable = true;
  # ...
  opts.showmode = false;
  plugins.noice = {
    enable = true;
    routes = [
      {
        view = "notify";
        filter = { event = "msg_showmode"; };
      }
    ];
  };
};
```
The `opts.showmode = false;` setting is needed to prevent notifications every time you change into visual mode, visual line mode, etc.
(https://github.com/folke/noice.nvim/issues/425#issuecomment-1605759747)

I'd still rather use the lualine solution, as I'd like some indicator to stay on screen as long as I'm recording a macro, not just a notification when I start a recording. So if anyone knows how to do that with NixVim, please let me know!

**Comment by MattSturgeon** [comment, marked as answer]
> I'd still rather use the lualine solution, as I'd like some indicator to stay on screen as long as I'm recording a macro, not just a notification when I start a recording. So if anyone knows how to do that with NixVim, please let me know!

This is made more complex because we haven't migrated lualine to use rfc42 style `settings` options. If it supported that, you could translate the lua config you linked as follows:

```nix
plugins.lualine.settings = {
  sections = {
    lualine_x = [
      {
        "__unkeyed.1".__raw = "require('noice').api.statusline.mode.get";
        cond.__raw = "require('noice').api.statusline.mode.has";
        color = { fg = "#ff9e64"; };
      }
    ];
  };
};
```

Because we don't (yet) have the plugin setup that way, you'll have to use the [plugins.lualine.sections.lualine_x](https://nix-community.github.io/nixvim/plugins/lualine/sections.html#pluginslualinesectionslualine_x) option instead, which is a list of `string or submodule`. Unfortunately the submodule options aren't documented, but you can [view the source here](https://github.com/nix-community/nixvim/blob/70088f6f8945023115f090f26764e5a71ae87a91/plugins/statuslines/lualine.nix#L27-L91).

I think something like this would work with our current implementation:
```nix
plugins.lualine.sections.lualine_x = [
  {
    name.__raw = "require('noice').api.statusline.mode.get";
    color = { fg = "#ff9e64"; };
    extraConfig.cond.__raw = "require('noice').api.statusline.mode.has";
  }
];
```

**Reply by lplawlor** [reply]
Yep, that works perfectly. Thank you!
